File: src/services/WorkspaceService.cpp
#include "WorkspaceService.hpp"
#include "core/HyprlandClient.hpp"

#include <print>

WorkspaceService::WorkspaceService(HyprlandListener& listener) : listener(listener), subscription_id(0) {
    this->subscription_id = this->listener.subscribe([this](const HyprlandEvent& event) { this->onEvent(event); });

    fetchInitialWorkspaces();
}

WorkspaceService::~WorkspaceService() {
    this->listener.unsubscribe(this->subscription_id);
}

void WorkspaceService::onEvent(const HyprlandEvent& event) {
    if (event.type == "workspace") {
        this->current_workspace = event.data;
        std::println("[WorkspaceService] State updated -> {}", this->current_workspace);
    }
    else if (event.type == "createworkspace") {
        int new_id = all_workspaces.size() + 1;
        all_workspaces.push_back(Workspace{.id = new_id, .name = event.data});
        std::println("[WorkspaceService] Workspace created: {0}. Total: {1}", event.data, all_workspaces.size());
    }
    else if (event.type == "destroyworkspace") {
        std::string destroyed_name = event.data;
        std::erase_if(all_workspaces, [&](const Workspace& ws) { return ws.name == destroyed_name; });
        std::println("[WorkspaceService] Workspace destroyed: {0}. Total: {1}", destroyed_name, all_workspaces.size());
    }
}

void WorkspaceService::fetchInitialWorkspaces() {
    std::println("[WorkspaceService] Fetching initial workspace state...");

    auto workspaces = HyprlandClient::getWorkspaces();

    if (!workspaces.has_value()) {
        std::println(stderr, "[WorkspaceService] Error: {}", workspaces.error());
        return;
    }

    if (workspaces->is_array()) {
        for (const auto& item : *workspaces) {
            all_workspaces.push_back(Workspace{.id = item["id"], .name = item["name"]});
        }
    }

    std::println(stderr, "[WorkspaceService] Inital state loaded. Workspaces: ");
    for (const auto& ws : all_workspaces) {
        std::println("ID: {0} | Name: {1}", ws.id, ws.name);
    }
}

File: src/services/workspace_service.rs
use std::sync::{Arc, Mutex, MutexGuard};

use serde::Deserialize;

use crate::{
  hyprland::{hyprland_event::HyprlandEvent, hyprland_ipc::HyprlandIpc},
  shared::LockError,
};

#[derive(Deserialize, Debug, Clone)]
pub struct Workspace {
  pub id: i32,
  pub name: String,
}

#[derive(Default, Debug)]
pub struct WorkspaceState {
  active_workspace: String,
  all_workspaces: Vec<Workspace>,
}

pub struct WorkspaceService {
  state: Arc<Mutex<WorkspaceState>>,
}

impl WorkspaceService {
  pub fn new(listener: Arc<Mutex<HyprlandIpc>>) -> Self {
    let state = Arc::new(Mutex::new(WorkspaceState::new()));

    let state_clone = Arc::clone(&state);
    if let Ok(mut listen) = listener.lock() {
      listen.subscribe(Box::new(move |event| {
        if let Ok(mut st) = state_clone.lock() {
          st.on_event(event);
        }
      }));
    }

    Self { state }
  }

  fn get_state<P, R>(&self, getter: P) -> Result<R, LockError>
  where
    P: FnOnce(MutexGuard<WorkspaceState>) -> R,
  {
    return match self.state.lock() {
      Ok(guard) => Ok(getter(guard)),
      Err(e) => Err(format!("Failed to acquire lock: {}", e)),
    };
  }

  pub fn get_active_workspace(&self) -> Result<String, LockError> {
    self.get_state(|guard| guard.active_workspace.clone())
  }

  pub fn get_all_workspaces(&self) -> Result<Vec<Workspace>, LockError> {
    self.get_state(|guard| guard.all_workspaces.clone())
  }
}

impl WorkspaceState {
  pub fn new() -> WorkspaceState {
    let mut initial_state = WorkspaceState::default();

    println!("[WorkspaceService] Fetching workspaces...");

    match HyprlandIpc::send_command("j/workspaces") {
      Ok(response) => {
        initial_state.all_workspaces = serde_json::from_str(&response).unwrap_or_default();
        println!("[WorkspaceService] Intial {initial_state:?}");
      },
      Err(e) => {
        eprintln!("[WorkspaceService] Failed to get initial workspaces {e}");
      },
    }

    initial_state
  }

  fn on_event(&mut self, event: HyprlandEvent) {
    match event.name.as_str() {
      "workspace" => {
        println!("[WorkspaceService] Active workspace {}", event.data);
        self.active_workspace = event.data;
      },
      "createworkspacev2" => {
        if let Some((id_str, name)) = event.data.split_once(',') {
          if let Ok(id) = id_str.parse::<i32>() {
            self.all_workspaces.push(Workspace { id, name: name.to_string() });
          }
        }
      },
      "destroyworkspacev2" => {
        if let Some((id_str, _)) = event.data.split_once(',') {
          if let Ok(id) = id_str.parse::<i32>() {
            self.all_workspaces.retain(|ws| ws.id != id);
          }
        }
      },
      _ => (),
    }
  }
}
